### PR TITLE
feat: support for generating testcontainers-go tests for DB drivers

### PR DESCRIPTION
**Pull request by mdelapenya**
By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

## Problem/Feature

Testing is a mayor concern nowadays, so this PR extends the code generation tool to include integration tests for the DB layer.

It uses [Testcontainers for Go](https://golang.testcontainers.org/) (spoiler alert: core maintainer here :wave:) to run the test. Given the project already uses Docker Compose for running the project in dev mode, I think it's a good idea to have real services backing the database layer at test time.

Sample ouput of the tests:
```
go test -timeout 600s -coverprofile=/var/folders/_j/nhbgdck523n3008dd3zlsm5m0000gn/T/vscode-goOsS6fE/go-code-cover testfoo/internal/database -count=1 -v

2024/07/03 17:49:31 github.com/testcontainers/testcontainers-go - Connected to docker:
  Server Version: 78+testcontainerscloud (via Testcontainers Desktop 1.15.0)
  API Version: 1.43
  Operating System: Ubuntu 20.04 LTS
  Total Memory: 7407 MB
  Resolved Docker Host: tcp://127.0.0.1:61939
  Resolved Docker Socket Path: /var/run/docker.sock
  Test SessionID: 044df7af03224e9fe70ad6713b479a3170777bb29ce94d5f4edf802c5afa2aa0
  Test ProcessID: 6d5f6261-a0e1-4e35-b0ba-9c70f33d2c91
2024/07/03 17:49:31 🐳 Creating container for image testcontainers/ryuk:0.7.0
2024/07/03 17:49:31 ✅ Container created: 160fe2de6307
2024/07/03 17:49:31 🐳 Starting container: 160fe2de6307
2024/07/03 17:49:31 ✅ Container started: 160fe2de6307
2024/07/03 17:49:31 🚧 Waiting for container id 160fe2de6307 image: testcontainers/ryuk:0.7.0. Waiting for: &{Port:8080/tcp timeout:<nil> PollInterval:100ms}
2024/07/03 17:49:31 🔔 Container is ready: 160fe2de6307
2024/07/03 17:49:31 🐳 Creating container for image mongo:latest
2024/07/03 17:49:31 ✅ Container created: 908e45f9aa93
2024/07/03 17:49:31 🐳 Starting container: 908e45f9aa93
2024/07/03 17:49:32 ✅ Container started: 908e45f9aa93
2024/07/03 17:49:32 🚧 Waiting for container id 908e45f9aa93 image: mongo:latest. Waiting for: &{timeout:<nil> deadline:<nil> Strategies:[0x140003f0480 0x1400007bf00]}
2024/07/03 17:49:32 🔔 Container is ready: 908e45f9aa93
=== RUN   TestNew
--- PASS: TestNew (0.00s)
=== RUN   TestHealth
--- PASS: TestHealth (0.03s)
PASS
coverage: 80.0% of statements
2024/07/03 17:49:32 🐳 Terminating container: 908e45f9aa93
2024/07/03 17:49:33 🚫 Container terminated: 908e45f9aa93
ok  	testfoo/internal/database	2.840s	coverage: 80.0% of statements
```

## Description of Changes: 

- Add templates for each supported driver: Mongo, Redis, MySQL, Postgres
- Use the existing Testcontainers for Go modules for such drivers: [Mongo](https://golang.testcontainers.org/modules/mongodb/), [Redis](https://golang.testcontainers.org/modules/redis/), [MySQL](https://golang.testcontainers.org/modules/mysql/), [Postgres](https://golang.testcontainers.org/modules/postgres/)
- Include templating mechanism and data structures to render the new templates
- Document the new capability
- Include a make goal to run the integration tests
- Include a Github action workflow for building an app for each DB driver, running the tests for each one

## Checklist

- [x] I have self-reviewed the changes being requested
- [x] I have updated the documentation (check issue ticket #218)
- [x] I have generated an app for each DB driver, and run its tests, all of them passing
- [x] I have added a GH workflow running the generating tests, demonstrating they are passing

### CODE REVIEW

**Comment by Ujstor** [comment]
@mdelapenya  Thanks for the PR, and sorry for the delay. We merged some other PRs first that included fixes and had priority. This is a very interesting one.

I think it would be better if we include the DB tests in the core blueprint and bind them to the DB driver flag. It makes more sense to me. The advanced flag is for more niche features. DB integration tests belong in the core.

The current workflow that you created is failing because in v.0.7.0, the core has a git flag that needs to be added: -g commit. After that, the workflow should pass.

Earlier, I saw that linter passed, but now it fails. There was no big change, only a refactor, and git was added into the matrix.

**Comment by mdelapenya** [comment]
> Earlier, I saw that linter passed, but now it fails. There was no big change, only a refactor, and git was added into the matrix.

@Ujstor thanks for your feedback! Indeed, the linter passed before and I'm fixing it. We released `v0.32.0` past week and it deprecated the `RunContainer` module APIs in favour of simply `Run`. Please see https://github.com/testcontainers/testcontainers-go/pull/2610

I'm working on the fix right now.

**Comment by mdelapenya** [comment]
> The advanced flag is for more niche features. DB integration tests belong in the core.

I protected the addition of the tests with the advanced flag in the case the team was not sure about bundling testcontainers into the projects as part of the core. I'm glad you see it adding them as part of the core as something positive. Thanks!

**Comment by mdelapenya** [comment]
@Ujstor PR is updated, including the tests in the core. I hope everything is correct. PLMK if anything needs to be changed, thanks!

File: .github/workflows/testcontainers.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      matrix:
+        driver:
+          [mysql, postgres, mongo, redis]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Setup Go
+        uses: actions/setup-go@v5
+        with:
+          go-version: '1.22.2'
```
Comment (Ujstor) [inline .github/workflows/testcontainers.yml:19]: nit '1.22.x' will use the latest release.

File: docs/docs/advanced-flag/testcontainers.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+## Running the tests
+
+Go to the `internal/database` directory and run the following command:
+
+```bash
+go test -v
+```
+
+Testcontainers automatically downloads the required Docker images and start the containers. The tests run against the containers, and once the tests are done, the containers are stopped and removed. For further information, refer to the [official documentation](https://golang.testcontainers.org/).
```
Comment (Ujstor) [inline docs/docs/advanced-flag/testcontainers.md:66]: You forgot to remove the old docs?

**Review by Ujstor — CHANGES_REQUESTED** [review]
Just two small things, and after that, we are ready to merge.

File: docs/docs/advanced-flag/testcontainers.md (hunk 2 above)
Comment (mdelapenya) [reply docs/docs/advanced-flag/testcontainers.md:66]: Oh my bad 🤦🏻

I'm on PTO for two weeks, feel free to remove that file in a commit of you want

File: .github/workflows/testcontainers.yml (hunk 1 above)
Comment (mdelapenya) [inline .github/workflows/testcontainers.yml:19]:
```suggestion
          go-version: '1.22.x'
```

**Review by Ujstor — APPROVED** [review]
LGTM

**Pull request merged**